merino-cache/src/redis/mod.rs

```diff
@@ -125,6 +130,99 @@ where
+        let mut connection = self.redis_connection.clone();
+
+        tracing::trace!("🔒Checking key: {:?}", &pending_key);
+        let lock = redis::pipe()
+            .add_command(redis::Cmd::get(&pending_key))
+            .query_async::<redis::aio::ConnectionManager, String>(&mut connection)
+            .instrument(tracing::info_span!("getting-cache-pending", %pending_key))
+            .await;
+
+        let locked = !lock.unwrap_or_default().is_empty();
+        tracing::trace!("🔒Is Pending {:?} {:?}", &pending_key, &locked);
+        Ok(locked)
+    }
+
+    /// Check to see if the lock we have matches the lock we made.
+    async fn check_pending(&self, key: &str, lock: &str) -> bool {
+        let mut conn = self.redis_connection.clone();
+        let pending_key = Self::pending_key(key);
+        let locked = match redis::cmd("GET")
+            .arg(&pending_key)
+            .query_async::<redis::aio::ConnectionManager, String>(&mut conn)
+            .await
+        {
+            Ok(set_lock) => {
+                tracing::trace!(
+                    "🔒Checking key: {:?} {:?} =? {:?}",
+                    &pending_key,
+                    set_lock,
+                    lock
+                );
+                set_lock == lock
+            }
+            Err(e) => {
+                tracing::warn!("🔒Could not get lock. Gone? {:?} {:?}", pending_key, e);
+                false
+            }
+        };
+        tracing::trace!("🔒Check Pending: {:?} {:?}", pending_key, locked);
+        locked
+    }
+
+    /// Generate a lock, this returns a unique Lock value string to ensure that
+    /// only the thread with the most recent "lock" can write to this key.
+    async fn set_pending(&self, key: &str) -> Result<String, SuggestError> {
+        let mut connection = self.redis_connection.clone();
+        let pending_key = Self::pending_key(key);
+        let lock = Uuid::new_v4().to_simple().to_string();
+        tracing::trace!("🔒Setting lock for {:?} to {:?}", &pending_key, &lock);
+        redis::pipe()
+            .add_command(redis::Cmd::set(&pending_key, &lock))
+            .add_command(redis::Cmd::expire(
+                &pending_key,
+                self.default_lock_timeout.as_secs().try_into().unwrap_or(3),
+            ))
+            .query_async(&mut connection)
+            .await
+            .map_err(|e| {
+                tracing::error!("🔒lock error: {:?}", e);
+                SuggestError::Internal(e.into())
+            })?;
+        Ok(lock)
+    }
+
+    /// Delete a lock (fails if the lock value does not match.)
+    async fn del_pending(&self, key: &str, lock: &str) -> Result<(), SuggestError> {
+        let mut connection = self.redis_connection.clone();
+        let pending_key = Self::pending_key(key);
+        if self.check_pending(key, lock).await {
+            tracing::trace!("🔒Removing lock for {:?}", &pending_key);
+            redis::cmd("DEL")
```

Comment:
Instead of having separate check and delete commands (which could be racey), you can use a [`Script`](https://docs.rs/redis/0.21.0/redis/struct.Script.html) command. That lets you send some Lua that will check if the value is the expected UUID, and then delete it, otherwise error out. This is all done automically.

Reply:
Ah, cool! TIL.

Now I just have to learn Lua quick.

I'll probably also update the write function to only write if the lock matches as well, because there's a bit of a potential race in there too.

Reply:
Yeah, Redis has this [pattern](https://redis.io/commands/set#patterns) to showcase the distributed locking on a single Redis node.